src/models/model.cpp

```diff
@@ -274,12 +274,21 @@ int32_t Tokenizer::TokenToTokenId(const char* token) const {
 }
 
 DeviceInterface* SetProviderSessionOptions(OrtSessionOptions& session_options,
+                                           const std::vector<std::string>& providers,
                                            const std::vector<Config::ProviderOptions>& provider_options_list,
                                            bool is_primary_session_options,
                                            bool disable_graph_capture) {
   DeviceInterface* p_device{};
 
-  for (auto& provider_options : provider_options_list) {
+  for (auto& provider : providers) {
+    auto provider_options_it = std::find_if(provider_options_list.begin(), provider_options_list.end(),
+                                            [&provider](const Config::ProviderOptions& po) { return po.name == provider; });
+
+    if (provider_options_it == provider_options_list.end()) {
```

Comment:
Wouldn't a missing provider_options just mean to use defaults for that provider?

Reply:
Any provider added to the providers list must also exist in the provider options since that was `SetProviderOption` does. We wouldn't expect there to be a provider string in providers that isn't in the provider options.

Reply:
Ah, I would, as previously the provider_options was the list of providers + options. But now there's two lists. So you could have a provider with only default options. I don't think it affects the outside world currently, as our API always goes through `SetProviderOption` like you said. So up to you.

Reply:
Will leave it as is for now. It won't ever hit this code (I could change to an assert) since the providers and provider options should always be in sync (i.e. provider options would always be a superset of providers).

We can update this code when we decide to change this assumption.